compiler_gym/datasets/tar_dataset.py

```diff
@@ -0,0 +1,204 @@
+                shutil.rmtree(self.site_data_path / "contents", ignore_errors=True)
+
+                arc.extractall(str(self.site_data_path / "contents"))
+
+            self._tar_extracted_marker.touch()
+
+        if self.strip_prefix and not self.dataset_root.is_dir():
+            raise FileNotFoundError(
+                f"Directory prefix '{self.strip_prefix}' not found in dataset '{self.name}'"
+            )
+
+
+class TarDatasetWithManifest(TarDataset):
+    """A tarball-based dataset that has a manifest file which lists the URIs.
+
+    The idea is to allow the list of benchmark URIs to be enumerated in a more
+    lightweight manner than downloading and unpacking the entire dataset. It
+    does this by downloading a "manifest", which is a list of benchmark names,
+    and only downloads the actual tarball containing the benchmarks when it is
+    needed.
+
+    The manifest is assumed to be correct and is not validated.
+    """
+
+    def __init__(
+        self,
+        manifest_urls: List[str],
+        manifest_sha256: str,
+        manifest_compression: str = "bz2",
+        **dataset_args,
+    ):
+        """Constructor.
+
+        :param manifest_urls: A list of redundant URLS to download the
+            compressed text file containing a list of benchmark URI suffixes,
+            one per line.
+
+        :param manifest_sha256: The sha256 checksum of the compressed manifest
+            file.
+
+        :param manifest_compression: The manifest compression type. One of
+            {"bz2", "gz"}.
+
+        :param dataset_args: See :meth:`TarDataset.__init__()
+            <compiler_gym.datasets.TarDataset.__init__>`.
+        """
+        super().__init__(**dataset_args)
+        self.manifest_urls = manifest_urls
+        self.manifest_sha256 = manifest_sha256
+        self.manifest_compression = manifest_compression
+        self._manifest_path = self.site_data_path / f"manifest-{manifest_sha256}.txt"
+
+        self._manifest_lock = Lock()
+        self._manifest_lockfile = self.site_data_path / "manifest.LOCK"
+
+    def _read_manifest_file(self) -> List[str]:
+        """Read the manifest file from disk.
+
+        Does not check that the manifest file exists.
+        """
+        with open(self._manifest_path, encoding="utf-8") as f:
+            lines = f.read().rstrip().split("\n")
+            uris = [f"{self.name}/{line}" for line in lines]
+            self.logger.debug("Read %s manifest, %d entries", self.name, len(uris))
+        return uris
+
+    @memoized_property
+    def _benchmark_uris(self) -> List[str]:
+        """Fetch or download the URI list."""
+        if self._manifest_path.is_file():
+            return self._read_manifest_file()
+
+        with self._manifest_lock:
+            with fasteners.InterProcessLock(self._manifest_lockfile):
+                # Now that we have acquired the lock, repeat the check.
+                if self._manifest_path.is_file():
```

Comment:
Consider two threads *X* and *Y* that both call into this function. *X* gets the lock first, causing *Y* to wait. *X* downloads the file and releases the lock. *Y*, having acquired the lock, should now check to see whether the file needs downloading, as otherwise it would be redundantly repeating the work that *X* just completed.

Hope that makes sense :)